packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +531,820 @@ class ShellRoute extends RouteBase {
+        _branchStates.add(currentBranchState);
+      } else {
+        _branchStates.add(existingStates.firstWhereOrNull(
+                (_StatefulShellBranchState e) => e.branch == branch) ??
+            _StatefulShellBranchState._(branch));
+      }
+    }
+
+    // Remove any obsolete cached Navigators
+    final Set<Key> validKeys =
+        _route.branches.map((StatefulShellBranch e) => e.navigatorKey).toSet();
+    _navigatorCache.removeWhere((Key key, _) => !validKeys.contains(key));
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _updateCurrentBranchStateFromWidget();
+  }
+
+  @override
+  void didUpdateWidget(covariant _StatefulNavigationShell oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _updateCurrentBranchStateFromWidget();
+    _preloadBranches();
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _preloadBranches();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> children = _route.branches
+        .map((StatefulShellBranch branch) => _BranchNavigatorProxy(
+            branch: branch, navigatorForBranch: _navigatorForBranch))
+        .toList();
+    return widget._containerBuilder(context, shellState, children);
+  }
+}
+
+typedef _NavigatorForBranch = Widget? Function(StatefulShellBranch);
+
+/// Widget that serves as the proxy for a branch Navigator Widget, which
+/// possibly hasn't been created yet.
+class _BranchNavigatorProxy extends StatelessWidget {
+  const _BranchNavigatorProxy({
+    required this.branch,
+    required this.navigatorForBranch,
+  });
+
+  final StatefulShellBranch branch;
+  final _NavigatorForBranch navigatorForBranch;
+
+  @override
+  Widget build(BuildContext context) {
+    return navigatorForBranch(branch) ?? const SizedBox.shrink();
+  }
+}
+
+/// Default implementation of a container widget for the [Navigator]s of the
+/// route branches. This implementation uses an [IndexedStack] as a container.
+class _IndexedStackedRouteBranchContainer extends StatelessWidget {
+  const _IndexedStackedRouteBranchContainer(
+      {required this.currentIndex, required this.children});
+
+  final int currentIndex;
+
+  final List<Widget> children;
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> stackItems = children
+        .mapIndexed((int index, Widget child) =>
+            _buildRouteBranchContainer(context, currentIndex == index, child))
+        .toList();
+
+    return IndexedStack(index: currentIndex, children: stackItems);
+  }
+
+  Widget _buildRouteBranchContainer(
+      BuildContext context, bool isActive, Widget child) {
+    return Offstage(
+      offstage: !isActive,
+      child: TickerMode(
+        enabled: isActive,
+        child: child,
+      ),
+    );
+  }
+}
+
+typedef _BranchSwitcher = void Function(int);
+
+/// Internal [StatefulShellRouteState] implementation.
+class _StatefulShellRouteState implements StatefulShellRouteState {
```

Comment:
This seems redundant, I will just keep the implementation in StatefulShellRouteState and make it a concrete class